Initial version of a json schema for the service specification

JordiPolo opened the pull request:
As discussed in #1083, adding a JSON Schema file.

This is NOT ready to merge. This file currently only implements [ADR-02](https://github.com/spotify/backstage/blob/master/docs/architecture-decisions/adr002-default-catalog-file-format.md) but needs to be expanded to support [ADR-05](https://github.com/spotify/backstage/blob/master/docs/architecture-decisions/adr005-catalog-core-entities.md)

In this file, I've left `namespace`, `description`, `labels` and `annotations` as optional and everything else required. You can't add extra keys anywhere except directly inside the `metadata` object.


@freben  @Rugvip

freben commented on docs/service_specification.schema.json:
```diff
@@ -0,0 +1,117 @@
+        "apiVersion": {
+            "type": "string",
+            "description": "Version of the specification format for a particular file is written against.",
+            "default": "backstage.io/v1beta1",
```
apiVersion and kind must always be specified, explicitly, so no defaults

freben commented on docs/service_specification.schema.json:
```diff
@@ -0,0 +1,117 @@
+            "type": "string",
+            "description": "High level entity type being described, from the Backstage system model.",
+            "enum": [
+                "Compontent"
```
replace with pattern

freben commented on docs/service_specification.schema.json:
```diff
@@ -0,0 +1,117 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema",
+    "$id": "backstage.io/v1beta1",
```
are these meant to be stable over time? or can this be like `backstage.io/catalog/entity`?

JordiPolo replied:
`The $id keyword defines a URI for the schema, and the base URI that other URI references within the schema are resolved against.`

So I think probably this should point to something in backstage.io or maybe to the github url to the master branch for this file.

freben commented on docs/service_specification.schema.json:
```diff
@@ -0,0 +1,117 @@
+    "required": [
+        "apiVersion",
+        "kind",
+        "spec",
```
spec is optional

JordiPolo replied:
Removed although I find it weird that spec which is supposed to represent the entity is optional but metadata is not?

freben commented on docs/service_specification.schema.json:
```diff
@@ -0,0 +1,117 @@
+        },
+        "spec": {
+            "type": "object",
+            "description": "Actual specification data that describes the entity. TODO: shape depend on `kind`",
```
see the main review comment - distinct kinds will probably be separate schema files

freben commented on docs/service_specification.schema.json:
```diff
@@ -0,0 +1,117 @@
+            "properties": {
+                "name": {
+                    "type": "string",
+                    "pattern": "^[a-z0-9A-Z_\\-\\.]{1,63}$",
```
dot does not have to be escaped in brackets, and if you put the dash as the first or last character you can get away with not escaping that too :)

freben commented:
Regarding ADR005, we'll be heavily focusing on the Component kind for the first iteration. So don't worry too much about the other ones just yet.

freben commented:
Or maybe make a separate definitions schema e.g. for the entire metadata block that anybody can ref? https://json-schema.org/understanding-json-schema/structuring.html

freben commented:
Yeah so it looks like we could make a individual kind schema be `allOf` both a ref to the base schema as well as its own additions

https://json-schema.org/understanding-json-schema/structuring.html#id5

jcarres-mdsol commented:
I think probably it would be possible to have a schema that applies to everything.
json-schema has the idea of using a key value to discriminate between schemas. The one thing almost sure you will want to have is different versions in different files.

JordiPolo commented:
I think probably it would be possible to have a schema that applies to everything.
json-schema [can express](https://json-schema.org/understanding-json-schema/reference/conditionals.html) having or not properties depending on the value of other properties.
The one thing almost sure you will want to have is different versions in different files.

jcarres-mdsol commented on docs/service_specification.schema.json:
```diff
@@ -0,0 +1,117 @@
+            "description": "Version of the specification format for a particular file is written against.",
+            "default": "backstage.io/v1beta1",
+            "enum": [
+                "backstage.io/v1beta1"
```
I'll try to have all variations in this file first. That would save the software to deal with trying to figure out what file to use. If not possible then will do as you suggest and break it into base + each specific

JordiPolo commented:
@freben Moved the file, renamed it to alpha1, added owner and lifecycle and made `spec` structure dependent of `kind`. So I think this file can apply to everything, you just need to add new spec definitions when they come.

I do think different versions should be different files, else this will be ginormous and difficult to read.
